### PR TITLE
Added the Wolf

### PR DESCRIPTION
**Pull request by hpharmsen**
Wolf is er nu nog eentje maar kunnen er ook meer worden. 
Werkt een zoals SmartCow op zoek gaat naar gras maar dan op zoek naar andere creatures.

Mmm, ik bedenk me net dat meerde wolven elkaar waarschijnlijk gaan aanvallen. Todo item.

### CODE REVIEW

File: settings.yaml [hunk 1]
```diff
@@ -6,15 +6,16 @@ move_cost: 2
...
 start_num_creatures: 6
+start_num_wolves: 2
 start_grass_fraction: 0.3
 start_rock_fraction: 0.05
 start_water_fraction: 0.01
 grass_grow_per_turn: 3
-world_size: 60
-scale: 10
+world_size: 35
+scale: 14
```
Comment (DOsinga) [inline settings.yaml]: Let's not merge these changes unless we have a good reason

File: settings.yaml [hunk 2]
```diff
@@ -6,15 +6,16 @@ move_cost: 2
...
 
 view_distance: 2
 steps_per_episode: 1000
 
-layers: [24, 24]
+layers: [24, 16]
```
Comment (DOsinga) [inline settings.yaml]: Same

File: simplegrid/cow.py [hunk 3]
```diff
@@ -37,20 +38,21 @@ def to_observation(self, offset):
...
 
     def __init__(self, x, y, settings, energy=None):
         self.x = x
         self.y = y
         self.settings = settings
         self.energy = energy or settings.init_energy
-        self.id = AbstractCow.id_count
-        AbstractCow.id_count += 1
+        self.id = AbstractCreature.id_count
+        self.is_predator = 0
```
Comment (DOsinga) [inline simplegrid/cow.py]: What is this? Why is it an int?

File: simplegrid/cow.py [hunk 4]
```diff
@@ -61,7 +63,10 @@ def learn(self, reward, done):
 
     def draw(self, display):
         display.circle(
-            self.x, self.y, math.sqrt(min(0.64, 2 * self.energy / MAX_ENERGY)), self.__class__.COLOR
+            self.x,
+            self.y,
+            math.sqrt(min(0.64, 2 * self.energy / MAX_ENERGY)),
+            self.__class__.COLOR,
```
Comment (DOsinga) [inline simplegrid/cow.py]: It feels like there is a line length change here in the config that is not part of this PR

File: simplegrid/cow.py [hunk 5]
```diff
@@ -37,20 +38,21 @@ def to_observation(self, offset):
         return direction[0] + offset, direction[1] + offset
 
 
-class AbstractCow(abc.ABC):
+class AbstractCreature(abc.ABC):
```
Comment (DOsinga) [inline simplegrid/cow.py]: I'd suggest moving this to its own file, now that it is no longer, well, a cow

File: simplegrid/dqn_agent.py [hunk 6]
```diff
@@ -138,16 +138,19 @@ def save_weights(self, name):
...
+                    state = [float(x) for x in state]
+                    if next is not None:
+                        next_state = [float(x) for x in next_state]
+                    record = {
+                        'state': state,
+                        'action': int(action),
+                        'reward': float(reward),
+                        'next_state': next_state,
+                    }
+                    fout.write(json.dumps(record) + '\n')
```
Comment (DOsinga) [inline simplegrid/dqn_agent.py]: As discussed.

File: settings.yaml (hunk 1 above)
Comment (hpharmsen) [reply settings.yaml]: No reason indeed. I'll change them back.

File: simplegrid/cow.py (hunk 3 above)
Comment (hpharmsen) [reply simplegrid/cow.py]: Line should be deleted. IS_PREDATOR is now a property of the class.

File: simplegrid/cow.py (hunk 5 above)
Comment (hpharmsen) [reply simplegrid/cow.py]: Done

File: simplegrid/cow.py (hunk 4 above)
Comment (hpharmsen) [reply simplegrid/cow.py]: I set line length to 120. I think you do too.
But somehow in master it's 99.
I will try to change it again.

File: simplegrid/dqn_agent.py (hunk 6 above)
Comment (hpharmsen) [reply simplegrid/dqn_agent.py]: Please update to what you think is right.

File: simplegrid/abstractcreature.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+    def learn(self, reward, done):
+        pass
+
+    def draw(self, display):
+        display.circle(self.x, self.y, math.sqrt(min(0.64, 2 * self.energy / MAX_ENERGY)), self.__class__.COLOR)
+
+    def split(self):
+        new_creature = self.__class__(self.x, self.y, self.settings, self.energy / 2)
+        self.energy /= 2
```
[the remark this answers is not in the document]
Comment (hpharmsen) [reply simplegrid/abstractcreature.py]: Not necessarily. Why?

File: simplegrid/wolf.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+WOLF_MOVE_SPEED = 0.5
+
+
+class Wolf(AbstractCreature):
+
+    COLOR = (101, 67, 33)
+    IS_PREDATOR = True
+
+    def step(self, observation):
+        self.energy = MAX_ENERGY
```
[the remark this answers is not in the document]
Comment (hpharmsen) [reply simplegrid/wolf.py]: Is a bit ugly I agree. The thing is that in the current setup wolves should no die. So I refill their energy every step. Later we might have evolving wolves with their own energy bookkeeping.

File: simplegrid/world.py [hunk 9]
```diff
@@ -196,7 +214,16 @@ def process_action(self, creature, action):
             # Move
             self.set_cell(creature.x, creature.y, 0)
             x, y = self.apply_direction(action, creature.x, creature.y)
-            if self.cells[x, y] in (MapFeature.GRASS.index, MapFeature.EMPTY.index):
+            if creature.__class__.IS_PREDATOR:
```
[the remark this answers is not in the document]
Comment (hpharmsen) [reply simplegrid/world.py]: First I checked for the class name being 'Wolf' but comparing to a string didn't feel right. 
We have multiple herbivores so why not multiple predators. Especially when we start training the wolves.
Anyway, adding a class method is no problem.

File: simplegrid/world.py [hunk 10]
```diff
@@ -88,10 +94,14 @@ def get_observation(self, creature):
...
         rolled = np.roll(self.cells, (size_2 - creature.x, size_2 - creature.y), (0, 1))
         view_distance = self.settings.view_distance
-        return rolled[
-            size_2 - view_distance : size_2 + view_distance + 1,
-            size_2 - view_distance : size_2 + view_distance + 1,
+        observation = rolled[
+            size_2 - view_distance : size_2 + view_distance + 1, size_2 - view_distance : size_2 + view_distance + 1
         ]
+        for x in range(observation.shape[0]):
+            for y in range(observation.shape[0]):
```
[the remark this answers is not in the document]
Comment (hpharmsen) [reply simplegrid/world.py]: Oh, I thought I fixed that.
Will do.

File: simplegrid/world.py [hunk 11]
```diff
@@ -88,10 +94,14 @@ def get_observation(self, creature):
...
+                if observation[x, y] > 0:
+                    observation[x, y] = self.creatures[observation[x, y]].__class__.IS_PREDATOR and 2 or 1
```
[the remark this answers is not in the document]
Comment (hpharmsen) [reply simplegrid/world.py]: I made the change and used MapFeature.COW.value and MapFeature.WOLF.value but beware, we are talking preys and predators here. Not WOLFs and COWs per se.

File: simplegrid/wolf.py [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+        possible_actions = {a: random.random() * 0.1 for a in list(Action)[1:-1]}
+
+        for col in range(size):
+            x = col - view_distance
+            for row in range(size):
+                y = row - view_distance
+                dist = abs(x) + abs(y)
+                value = observation[col, row]
+                if 0 < dist <= view_distance:
+                    if value == MapFeature.ROCK.index:
```
[the remark this answers is not in the document]
Comment (hpharmsen) [reply simplegrid/wolf.py]: Definitely

File: simplegrid/world.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (hpharmsen) [reply simplegrid/world.py]: Using a static method is even cleaner I think..

File: simplegrid/wolf.py [hunk 13]
```diff
@@ -31,6 +31,8 @@ def step(self, observation):
                 if 0 < dist <= view_distance:
                     if value == MapFeature.ROCK.index:
                         reward = -1
+                    elif value == MapFeature.WATER.index:
```
Comment (DOsinga) [inline simplegrid/wolf.py]: Currently trying to go into the water doesn't kill the Wolf, so maybe the same value as for rock would be ok?

File: simplegrid/world.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (DOsinga) [reply simplegrid/world.py]: Sure. We could rename them in the MapFeature too to be less confusing

File: simplegrid/wolf.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (DOsinga) [reply simplegrid/wolf.py]: I thought the Wolf was immortal because it never loses energy, so we don't need this here.

File: simplegrid/dqn_agent.py (hunk 6 above)
Comment (DOsinga) [reply simplegrid/dqn_agent.py]: Should be ok if you merge in master

File: simplegrid/world.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (DOsinga) [reply simplegrid/world.py]: Let's make it so!

[the remark this answers is not in the document]
Comment (hpharmsen) [reply simplegrid/world.py]: ok!
